RE: Build an API endpoint for shelves

xlisachan opened the pull request:
Fixes #14873 
** Currently failing Travis, as it is dependent on the migration file in #15242 **

This pull request builds an API endpoint (`/api/v4/shelves/` - or should this be `/api/shelves`?) that represents a list of shelves to appear on the AMO homepage.

I'm not 100% sure about the value for `permission_classes` within `shelves/views.py`.

Please review when you have a moment. Thank you!

eviljeff commented on src/olympia/shelves/views.py:
```diff
@@ -0,0 +1,13 @@
+
+class ShelfViewSet(viewsets.ModelViewSet):
+    queryset = ShelfManagement.objects.all()
+    permission_classes = [
+        permissions.AllowAny
+    ]
```
```suggestion
    permission_classes = []
```

eviljeff commented on src/olympia/shelves/serializers.py:
```diff
@@ -0,0 +1,18 @@
+                  'footer_text', 'footer_pathname']
+
+
+class HomepageSerializer(serializers.ModelSerializer):
```
So, I don't think we actually need this Serializer at all - we won't be returning any data from it directly.  The `ShelfSerializer` has everything we need.

eviljeff commented on src/olympia/shelves/serializers.py:
```diff
@@ -0,0 +1,18 @@
+class ShelfSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Shelf
+        fields = ['id', 'title', 'endpoint', 'criteria',
```
frontend won't need `id`.

eviljeff replied:
And `endpoint` and `criteria` won't mean much either - they should be combined into a url field.  (url can be a `SerializerMethodField` that returns a full api url - similar to how the url is being built in the clean methods)

eviljeff commented on src/olympia/shelves/views.py:
```diff
@@ -0,0 +1,13 @@
+        permissions.AllowAny
+    ]
+
+    serializer_class = HomepageSerializer
```
`ShelfSerializer`

eviljeff commented on src/olympia/shelves/views.py:
```diff
@@ -0,0 +1,13 @@
+
+
+class ShelfViewSet(viewsets.ModelViewSet):
+    queryset = ShelfManagement.objects.all()
```
So, really here we want a queryset of Shelf, but filtered and ordered by some of the ShelfManagement properties.  Something like `Shelf.objects.filter(shelfmanagement__enabled=True).order_by('shelfmanagement__position')`

eviljeff commented on src/olympia/shelves/serializers.py:
```diff
@@ -0,0 +1,18 @@
+
+class HomepageSerializer(serializers.ModelSerializer):
+    shelf = ShelfSerializer(required=True)
+
```
The next issue will be to actually include the contents of the shelves in the response.

xlisachan commented:
This latest push includes the following updates:
- Removed `HomepageSerializer`
- Combined `endpoint` and `criteria` into a `url` field
- Updated `serializer_class` to `ShelfSerializer` and `queryset` in `ShelfViewSet`
- Revised `test_serializers.py`

Please review when you have a moment. Thank you!

eviljeff commented on src/olympia/shelves/serializers.py:
```diff
@@ -0,0 +1,26 @@
+    def get_url(self, obj):
+        baseUrl = settings.INTERNAL_SITE_URL
+        if obj.endpoint == 'search':
+            api = drf_reverse('v4:addon-search')
```
because this is in the API we want to always use the matching API version in the reverses too.  So strip off the `v4:` at the start of the url name and include `request=self.context.get('request')` as a kwarg to pass down the request so the api version can be resolved from that by DRF.

eviljeff commented on src/olympia/shelves/serializers.py:
```diff
@@ -0,0 +1,26 @@
+            api = drf_reverse('v4:addon-search')
+            url = baseUrl + api + obj.criteria
+        elif obj.endpoint == 'collections':
+            api = drf_reverse('v4:collection-addon-list', kwargs={
```
same here.

eviljeff commented on src/olympia/shelves/tests/test_serializers.py:
```diff
@@ -0,0 +1,27 @@
+        assert serializer.data == {
+            'title': 'Populâr themes',
+            'url': (settings.INTERNAL_SITE_URL +
+                    drf_reverse('v4:addon-search') +
```
use `reverse_ns` in API related tests - it handles the api versioning.

eviljeff commented on src/olympia/shelves/views.py:
```diff
@@ -0,0 +1,12 @@
+        shelfmanagement__enabled=True).order_by('shelfmanagement__position')
+    permission_classes = []
+
+    serializer_class = ShelfSerializer
```
can we have a test for the view too?  Doesn't need to do anything complicated - just check that it does indeed return a couple of Shelf instances, in the correct order, and without disabled shelves.

xlisachan replied:
I added a `test_views.py`, but I'm not 100% sure about it. I got a `NoReverseMatch` error stating `Reverse for 'addon-search' not found. 'addon-search' is not a valid view function or pattern name` when I tried to use `ShelfSerializer(instance=shelf_a).data`, so I wrote lines 64-68 instead. Please take a look when you have a moment.

eviljeff replied:
yes, that's fine - ShelfSerializer now relies on a request being passed in to resolve the api url correctly so `ShelfSerializer(instance=shelf_a).data` won't work

xlisachan commented:
Thanks for the feedback! This latest push includes the following revisions:
- Updated `serializers.py` to match the API version in the reverse
- Use `reverse_ns` in `test_serializers.py` (I think I have to update `test_forms.py` 👀 as I used `drf_reverse` in there?)
- Added `test_views.py` - I got a `NoReverseMatch` error stating `Reverse for 'addon-search' not found. 'addon-search' is not a valid view function or pattern name` when I tried to use `ShelfSerializer(instance=shelf_a).data`. I'm not sure if lines 64-68 are acceptable.

Please review when you have a moment. Thank you!

eviljeff commented on src/olympia/shelves/tests/test_views.py:
```diff
@@ -0,0 +1,68 @@
+
+class TestShelfViewSet(TestCase):
+    def setUp(self):
+        self.url = reverse_ns('shelves-list', api_version='v4')
```
do you need to specify the `api_version` here?  (it should default to 'v5' which should work okay too)

xlisachan replied:
I'll remove `api_version` as it did work without it. I had made the endpoint for the shelves API `/api/v4/shelves/`, should it be changed?

eviljeff replied:
> I had made the endpoint for the shelves API /api/v4/shelves/, should it be changed?

where did you do that?

xlisachan replied:
Please see below.

xlisachan commented on src/olympia/api/urls.py:
```diff
@@ -32,6 +32,7 @@
     url(r'^', include('olympia.signing.urls')),
     url(r'^', include(amo_api_patterns)),
     url(r'^scanner/', include('olympia.scanners.api_urls')),
+    url(r'^shelves/', include('olympia.shelves.urls')),
```
I wasn't sure where to put this.
